File: view/decoder.py
# -*- coding: utf-8 -*-
decode_symbol = {
            11: '1',
            12: '2',
            13: '3',
            14: '4',
            15: '5',
            16: '6',
            21: 'Ⅰ',
            22: 'Ⅱ',
            23: 'Ⅲ',
            24: 'Ⅳ',
            25: 'Ⅴ',
            26: 'Ⅵ',
            31: 'A',
            32: 'B',
            33: 'C',
            34: 'D',
            35: 'E',
            36: 'F',
            41: '₽',
            42: '$',
            43: '€',
            44: '฿',
            # 44: '₿',
            45: '£',
            46: '¥',
            51: 'α',
            52: 'β',
            53: 'ω',
            54: 'δ',
            55: 'π',
            56: 'φ',
            61: '-',
            62: '+',
            63: '×',
            64: '÷',
            65: '=',
            66: '%'
        }


def symbol_for(value):
    """The glyph a cell value renders as (or '?' if unknown)."""
    return decode_symbol.get(value, '?')


def rule_segments(values):
    """Structured reading of a rule for the tooltip: a list of segments,
    each ('text', str) or ('cell', value). 'cell' segments render as a small
    coloured tile (the real board colour), not just the bare glyph.

    Markers: '^' same column, '<->' neighbouring columns, '...' left-of.
    Three integers instead mean three consecutive columns, in that order
    read either left-to-right or right-to-left.
    """
    a, b, c = values
    if b == '^':
        return [('cell', a), ('text', 'is in the same column as'),
                ('cell', c)]
    if b == '<->':
        return [('cell', a), ('text', 'and'), ('cell', c),
                ('text', 'are in neighbouring columns')]
    if b == '...':
        return [('cell', a), ('text', 'is somewhere to the left of'),
                ('cell', c)]
    return [('cell', a), ('cell', b), ('cell', c),
            ('text', 'fill three columns in a row, in this order —'),
            ('text', 'left-to-right or right-to-left')]


def describe_rule(values):
    """The same reading as rule_segments(), flattened to a plain string."""
    out = []
    for kind, val in rule_segments(values):
        out.append(symbol_for(val) if kind == 'cell' else val)
    return ' '.join(out)

File: phaser/src/model/decoder.ts
// Glyph each cell value renders as — mirror of view/decoder.py. Row (tens
// digit) picks the alphabet; value (units digit, 1..7) picks the glyph.
// Categories are chosen to be easy to "say in your head" and crisp when small:
// 1 numbers · 2 latin letters · 3 math signs · 4 greek letters ·
// 5 geometric shapes · 6 more latin letters · (7 musical notes — hidden 7×7).
export const DECODE_SYMBOL: Record<number, string> = {
  11: '1', 12: '2', 13: '3', 14: '4', 15: '5', 16: '6', 17: '7',
  21: 'A', 22: 'B', 23: 'C', 24: 'D', 25: 'E', 26: 'F', 27: 'G',
  31: '-', 32: '+', 33: '×', 34: '÷', 35: '=', 36: '%', 37: '∞',
  41: 'α', 42: 'β', 43: 'ω', 44: 'δ', 45: 'π', 46: 'φ', 47: 'σ',
  // geometric shapes — ▰ (parallelogram) stands in for a trapezoid, which the
  // bundled DejaVu Sans has no glyph for.
  51: '▲', 52: '■', 53: '●', 54: '◆', 55: '★', 56: '▰', 57: '♣',
  61: 'U', 62: 'V', 63: 'W', 64: 'X', 65: 'Y', 66: 'Z', 67: 'T',
  71: '♩', 72: '♪', 73: '♫', 74: '♬', 75: '♭', 76: '♮', 77: '♯',
};

export function symbolFor(value: number): string {
  return DECODE_SYMBOL[value] ?? '?';
}

/** Vertical anchor for a glyph. The black geometric-shape glyphs (row 5) sit
 *  visually low when centred on their text box, so anchor them slightly below
 *  centre (origin-y > 0.5) which nudges the drawn shape upward. */
export function glyphOriginY(value: number): number {
  return Math.floor(value / 10) === 5 ? 0.58 : 0.5;
}

export type Segment = { kind: 'text'; value: string } | { kind: 'cell'; value: number };

/** Structured reading of a rule for display: text spans + coloured cell tiles. */
export function ruleSegments(values: Array<number | string>): Segment[] {
  const [a, b, c] = values;
  if (b === '^') {
    return [
      { kind: 'cell', value: a as number },
      { kind: 'text', value: 'same column as' },
      { kind: 'cell', value: c as number },
    ];
  }
  if (b === '<->') {
    return [
      { kind: 'cell', value: a as number },
      { kind: 'text', value: '&' },
      { kind: 'cell', value: c as number },
      { kind: 'text', value: 'neighbour columns' },
    ];
  }
  if (b === '...') {
    return [
      { kind: 'cell', value: a as number },
      { kind: 'text', value: 'left of' },
      { kind: 'cell', value: c as number },
    ];
  }
  // three-in-a-row
  return [
    { kind: 'cell', value: a as number },
    { kind: 'cell', value: b as number },
    { kind: 'cell', value: c as number },
    { kind: 'text', value: '3 in a row (either direction)' },
  ];
}
